src/com/amazon/ion/impl/ResizingPipedInputStream.java

```diff
@@ -0,0 +1,557 @@
+     *                    available. When false, all buffered bytes will be available to read.
+     */
+    ResizingPipedInputStream(final int initialBufferSize, final int maximumBufferSize, final boolean useBoundary) {
+        if (initialBufferSize < 1) {
+            throw new IllegalArgumentException("Initial buffer size must be at least 1.");
+        }
+        if (maximumBufferSize < initialBufferSize) {
+            throw new IllegalArgumentException("Maximum buffer size cannot be less than the initial buffer size.");
+        }
+        this.initialBufferSize = initialBufferSize;
+        this.maximumBufferSize = maximumBufferSize;
+        this.capacity = initialBufferSize;
+        buffer = new byte[initialBufferSize];
+        byteBuffer = ByteBuffer.wrap(buffer, 0, capacity);
+        this.useBoundary = useBoundary;
+    }
+
+    /**
+     * Moves all buffered (but not yet read) bytes from 'buffer' to the destination buffer. In total, 'size' bytes
+     * will be moved.
+     * @param destinationBuffer the destination buffer, which may be 'buffer' itself or a new buffer.
+     */
+    private void moveBytesToStartOfBuffer(byte[] destinationBuffer) {
+        if (size > 0) {
+            System.arraycopy(buffer, readIndex, destinationBuffer, 0, size);
+        }
+        if (readIndex > 0) {
+            notificationConsumer.bytesConsolidatedToStartOfBuffer(readIndex);
+        }
+        readIndex = 0;
+        boundary = available;
+        writeIndex = size;
+    }
+
+    /**
+     * @return the number of bytes that can be written at the end of the buffer.
+     */
+    private int freeSpaceAtEndOfBuffer() {
+        return capacity - writeIndex;
+    }
+
+    /**
+     * Ensures that there is at least 'minimumNumberOfBytesRequired' bytes of free space in the buffer, growing the
+     * buffer if necessary. May consolidate buffered bytes, performing an in-order copy and resetting indices
+     * such that the `readIndex` points to the same byte and the `writeIndex` is positioned after the last
+     * byte that is available to read.
+     * @param minimumNumberOfBytesRequired the minimum amount of free space that needs to be available for writing.
+     */
+    private void ensureSpaceInBuffer(int minimumNumberOfBytesRequired) {
+        if (size < 1 || freeSpaceAtEndOfBuffer() < minimumNumberOfBytesRequired) {
+            // Free up space by moving any unread bytes to the start of the buffer and resetting the indices.
+            moveBytesToStartOfBuffer(buffer);
+            int shortfall = minimumNumberOfBytesRequired - freeSpaceAtEndOfBuffer();
+            if (shortfall > 0) {
+                // There is not enough space in the buffer even though all available bytes have already been
+                // moved to the start of the buffer. Growth is required.
+                int amountToGrow = Math.max(initialBufferSize, shortfall);
```

Comment:
Using a growth factor instead of growing by `initialBufferSize` each time is something I'd like to experiment with, and I'd prefer to do it as a fast-follow to this PR as it requires all new performance/memory benchmarks.

> Here when `minimumNumberOfBytesRequired > initialBufferSize` unbalanced write/read sequences will lead to allocation and copying on every write.

That's true if you're writing significantly more than you're reading. In that case, users should heed the warning on the Javadoc for the `initialBufferSize` parameter: "should be chosen carefully such that growth is expected to occur rarely, if ever." Each read frees bytes in the buffer, so for normal cases that read the data that is written fairly frequently, the size of the buffer should stabilize. For the Ion use case, for example, we know that the buffer will be written with a single top-level value (plus preceding system value(s)), then that data will be read. The size of the buffer will stabilize to approximately the size of the largest top-level value in the stream.

Using a growth factor requires fewer allocations/copies to stabilize, but will usually be more wasteful of memory. That's why I want to study the difference in-depth after resolving the other feedback on this PR.

Reply:
https://github.com/amzn/ion-java/issues/375